apps/emqx_authn/test/emqx_authn_mysql_SUITE.erl

```diff
@@ -23,70 +24,396 @@
-                   <<"SELECT password_hash, salt FROM users where username = ",
-                        ?PH_USERNAME/binary, " LIMIT 1">>
-               },
-    {ok, _} = update_config([authentication], {create_authenticator, ?GLOBAL, Config}),
-
-    meck:expect(emqx_resource, query,
-        fun(_, {sql, _, [<<"good">>], _}) ->
-            {ok, [<<"password_hash">>, <<"salt">>], [[PasswordHash, Salt]]};
-           (_, {sql, _, _, _}) ->
-            {error, this_is_a_fictitious_reason}
-        end),
-
-    ClientInfo = #{zone => default,
-                   listener => 'tcp:default',
-                   protocol => mqtt,
-                   username => <<"good">>,
-                   password => Password},
-    ?assertEqual({ok, #{is_superuser => false}}, emqx_access_control:authenticate(ClientInfo)),
-
-    ClientInfo2 = ClientInfo#{username => <<"bad">>},
-    ?assertEqual({error, not_authorized}, emqx_access_control:authenticate(ClientInfo2)),
-    emqx_authn_test_lib:delete_config(<<"password-based:mysql">>),
-    ?AUTHN:delete_chain(?GLOBAL).
-
-update_config(Path, ConfigRequest) ->
-    emqx:update_config(Path, ConfigRequest, #{rawconf_with_defaults => true}).
+t_create(_Config) ->
+    AuthConfig = raw_mysql_auth_config(),
+
+    {ok, _} = emqx:update_config(
+                ?PATH,
+                {create_authenticator, ?GLOBAL, AuthConfig}),
+
+    {ok, [#{provider := emqx_authn_mysql}]} = emqx_authentication:list_authenticators(?GLOBAL).
+
+t_create_invalid(_Config) ->
+    AuthConfig = raw_mysql_auth_config(),
+
+    InvalidConfigs =
+        [
+         maps:without([server], AuthConfig),
+         AuthConfig#{server => <<"unknownhost:3333">>},
+         AuthConfig#{password => <<"wrongpass">>},
+         AuthConfig#{database => <<"wrongdatabase">>}
+        ],
+
+    lists:foreach(
+      fun(Config) ->
+              {error, _} = emqx:update_config(
+                             ?PATH,
+                             {create_authenticator, ?GLOBAL, Config}),
+
+              {ok, []} = emqx_authentication:list_authenticators(?GLOBAL)
+      end,
+      InvalidConfigs).
+
+t_authenticate(_Config) ->
+    ok = lists:foreach(
+           fun(Sample) ->
+                   ct:pal("test_user_auth sample: ~p", [Sample]),
+                   test_user_auth(Sample)
+           end,
+           user_seeds()).
+
+test_user_auth(#{credentials := Credentials0,
+                 config_params := SpecificConfgParams,
+                 result := Result}) ->
+    AuthConfig = maps:merge(raw_mysql_auth_config(), SpecificConfgParams),
+
+    {ok, _} = emqx:update_config(
+                ?PATH,
+                {create_authenticator, ?GLOBAL, AuthConfig}),
+
+    Credentials = Credentials0#{
+                    listener => 'tcp:default',
+                    protocol => mqtt
+                   },
+
+    ?assertEqual(Result, emqx_access_control:authenticate(Credentials)),
+
+    emqx_authn_test_lib:delete_authenticators(
```

Comment:
Curiosity: is this needed, since it is called in [`init_per_testcase`](https://github.com/emqx/emqx/pull/6269/files#diff-fbcce8baa41047f02fef239af566f87d43c56545586ef1e4d2a70b9b819fe010R42)?

Reply:
This is `test_user_auth` function, that is not a testcase by itself, but is run in a loop against the list of samples.

Reply:
Oops, you're right. Ignore. =)